### PR TITLE
Registration endpoint, sidebar, voter roll page

### PR DESCRIPTION
**Pull request by mikefranze**
Added registration endpoint that allows a voter to be submitted to election roll in registered state. Also added registration data to election roll db for the future. 

Added very basic sidebar, only appears if you have a role in the election. 

Added page to view voter rolls and approve, flag, invalidate them. Need to add logic to make these permission based.

### CODE REVIEW

**Review by scottPlusPlus — CHANGES_REQUESTED** [review]
1. Can we get a test that demonstrates the functionality?

2. I'm a little lost as to what "registration" information is in this case?  What type of data are we expecting?

File: backend/src/Models/ElectionRolls.ts
Comment (scottPlusPlus) [inline backend/src/Models/ElectionRolls.ts]: I'm assuming the option to pass multiple voter_ids is ONLY used when setting up the election?

If so would consider breaking that into a separate method on the interface.

File: frontend/src/components/EditElectionRoll.tsx
Comment (scottPlusPlus) [inline frontend/src/components/EditElectionRoll.tsx:15]: don't need to address for this diff but:

do we have (or want) an option for a voter to see whether or not they've been approved yet?  (and thus actually counted)?

File: frontend/src/components/EditElectionRoll.tsx
Comment (scottPlusPlus) [inline frontend/src/components/EditElectionRoll.tsx:20]: theoretically can combine these into a single "request", with a type (validate, flag, unflag, etc) that determines the URL.  Then elsewhere in the page don't need to check for 4 different responses or errors, just of the active "request"

File: frontend/src/components/EditElectionRoll.tsx
Comment (mikefranze) [reply frontend/src/components/EditElectionRoll.tsx:15]: Yes I think so. When the voter goes to the election home page it currently says if they've voted, we can add additional info there.

File: backend/src/Models/ElectionRolls.ts
Comment (mikefranze) [reply backend/src/Models/ElectionRolls.ts]: Yea we really only do multiple when they're added in batch by the election host when setting up. Adding a separate method would be better for adding just one. Or I can refactor to input an array of the election rolls and move the mapping of voter ids to the controller.

**Comment by mikefranze** [comment]
> 1. Can we get a test that demonstrates the functionality?
> 2. I'm a little lost as to what "registration" information is in this case?  What type of data are we expecting?

Yea I'll add a test. The registration is mostly just a placeholder right now. I'm thinking the election settings would hold info on the format of the registration data, like Name, Address, Zip Code, and API to send the data to like the registered voter database. Then the users would be prompted to enter Name, Address, etc and the data would be stored in the registration field. The credentialer would also be able to view this registration info and approve.

**Comment by mikefranze** [comment]
Tests added

File: backend/src/Models/ElectionRolls.ts
Comment (mikefranze) [reply backend/src/Models/ElectionRolls.ts]: Updated to just pass in list of voter rolls.

File: backend/src/test/testInputs.ts
Comment (scottPlusPlus) [inline backend/src/test/testInputs.ts:54]: is this used somewhere?

**Review by scottPlusPlus — APPROVED** [review]
thanks for adding the test.

eventually will want to show that votes that are not approved will not count, and that an official could approve a registration before of after the user votes.  and ideally voter will have way to see if they are approved yet or not.

but don't want to hold a diff hostage because not everything is sorted yet, we can add eventually.

**Comment by mikefranze** [comment]
> thanks for adding the test.
> 
> eventually will want to show that votes that are not approved will not count, and that an official could approve a registration before of after the user votes. and ideally voter will have way to see if they are approved yet or not.
> 
> but don't want to hold a diff hostage because not everything is sorted yet, we can add eventually.

Correct, and also need to add that actual logic lol. But I think it would make sense to wait until we do some general refactoring first.

File: backend/src/Models/ElectionRolls.ts
Comment (ArendPeter) [inline backend/src/Models/ElectionRolls.ts:111]: This is outside the scope of this PR, but we should probably think about how to handle concurrent modifications down the line

Let's say there are currently 3 events in the history, and 2 clients try adding a 4th event at the same time. If the flow happened as follows, the second client could overwrite the addition on the first client added

Example
```
Client1: Downloads history (w/ 3 elements) 
Client2: Downloads history (same 3 elements)
Client1: Locally adds 4th event
Client2: Locally adds 4th event
Client1: Uploads their modified history
Client2: Uploads their modified history (thus overwriting client1's addition)
```

I've used optimistic locking ([explainer vid](https://www.youtube.com/watch?v=VxGKvqHhU5c)) for this in the past, but we can discuss other options

File: backend/src/Controllers/voterRolls.controller.ts
Comment (ArendPeter) [inline backend/src/Controllers/voterRolls.controller.ts:52]: How we store the history list could be worth more discussion. 

Currently looks like we're storing a separate history list for each user in the roll list (1 list per user/election)? Here's a list of some other options

- 1 list for everything (a global history)
- 1 list per election
- 1 list per election server (for the future when our architecture spreads the load across multiple servers)
- 1 list per ballot
- 1 list per voter roll (current approach)

I haven't decided which approach I like best yet. "1 list per election" will make the auditing process cleaner (otherwise we're combing a bunch of smaller lists), but it also makes concurrency issues more likely. "1 list per election tenant" will be ideal for the hashing, and "1 list per voter roll" approach is ideal for voters looking up their ballot info

File: backend/src/Controllers/voterRolls.controller.ts
Comment (ArendPeter) [inline backend/src/Controllers/voterRolls.controller.ts:70]: In the future I could see a bulkRegisterVoter being handy (for example, if IPO wanted to use this API to register their full voter list)

**Review by ArendPeter — APPROVED** [review]
Looks great! All of my comments are more for future PRs so no changes needed

File: backend/src/Controllers/voterRolls.controller.ts
Comment (scottPlusPlus) [reply backend/src/Controllers/voterRolls.controller.ts:52]: Could you elaborate what you mean that "1 list per election will make auditing process cleaner?"

I'm imagining the two most common questions would be:
* is this voter approved
* why is this voter approved / not approved

which seem to be served by history per voter per election
(I'm also seeing "1 history per voter per election" the same as "1 list per ballot", but if there's a distinction here please LMK)

Comment (mikefranze) [reply backend/src/Controllers/voterRolls.controller.ts:52]: I think the use for 1 list per election for an auditor to easily grab all the data and see for example all the voters that were rejected. You might also want voters to be able to see this data (with any sensitive data removed). But it should be easy enough to query all the histories and concatenate them if needed. 

My concern with 1 list is that it makes your concurrent modifications concern more likely. 

Another approach is we have a separate table for histories and just have 1 row per modification, but that sounds like overkill.

Comment (ArendPeter) [reply backend/src/Controllers/voterRolls.controller.ts:52]: Sure thing

| Could you elaborate what you mean that "1 list per election will make auditing process cleaner?"

I was thinking the auditor would want to see a full log of everything that happened in the election in one place

| I'm also seeing "1 history per voter per election" the same as "1 list per ballot", but if there's a distinction here please LMK

That one is more a question of where we store the history. There are actions like "vote counted", or "vote rejected", that are initiated by an admin rather than the voter, so maybe they make more sense attached to the ballot rather than the voter? Then again, the registration history would exist before the ballot, so maybe roll is the better option between the 2?

Comment (mikefranze) [reply backend/src/Controllers/voterRolls.controller.ts:52]: I'm thinking voter roll history would have things like registration or changing state, ballot history would have things like if the ballot is invalidated or changed because the voter resubmitted or invalidated by an admin. I think we'll need to come up with a list of reasons why any of these would change/need history data and go from there.

Comment (scottPlusPlus) [reply backend/src/Controllers/voterRolls.controller.ts:52]: Ah, great thing to clarify:  are there reasons a vote would be rejected other than having to do with the voter registration?  Something about the ballot itself rather than about the voter?

What all information are we trying to capture in these histories?

For ballot, I assumed when the voter votes, and if they change their ballot.

For the registration, would want to see when the initial registration happened, when registration info is uploaded / changed (ex voter id etc)., as well as when approving / rejecting the voter.

What other events we need to capture?

File: backend/src/Controllers/voterRolls.controller.ts
Comment (mikefranze) [reply backend/src/Controllers/voterRolls.controller.ts:70]: We sort of have that with adding the voter id list. Registering marks them as registered which would then need approval. But if IPO is adding them then they would already be approved?

**Pull request merged**
